src/components/structural/AppMenu.tsx

```diff
@@ -1,21 +1,177 @@
+import { Link, LinkProps, useMatch, useResolvedPath } from 'react-router-dom'
+
+const defaultTestIdElement = 'menu-element'
+const defaultIconClasses = 'transition ease-in-out duration-200 cursor-pointer text-primary-500 hover:text-primary-400'
 
 /**
  * This component provides the structure for a
  * navigation bar. Items can be passed as children.
  */
-const AppMenu = (props: AppMenuProps) => {
+function AppMenu(props: AppMenuProps) {
     return (
-        <div className="p-4 w-full fixed top-0 left-0" data-testid="app-menu">
-            <div className="h-14 p-2 w-full bg-primary-100 rounded-3xl flex flex-col items-center justify-center shrink-0 shadow-xl shadow-primary-900/[0.1]">
-                <nav className="space-x-4">{props.children}</nav>
+        <div className="p-4 w-full fixed top-0 left-0 z-50" data-testid="app-menu">
+            <div className="h-14 w-full bg-primary-100 rounded-3xl flex flex-row justify-center shrink-0 shadow-xl shadow-primary-900/[0.1]">
+                <nav
+                    className="flex flex-row px-4 py-2 justify-start grow items-center space-x-2 overflow-x-scroll"
+                    data-testid="app-menu-primary-items"
+                >
+                    {props.leadingItem && (
+                        <>
+                            <MenuElement fromItem={props.leadingItem} />
+                            <div className="w-0"></div>
+                        </>
+                    )}
+                    <PrimaryMenu {...props} />
+                </nav>
+                <div className="flex flex-row px-4 py-2 justify-end grow-0 items-center space-x-4">
+                    {props.trailingItem && <SecondaryMenu {...props} />}
+                </div>
             </div>
         </div>
     )
 }
-
 export default AppMenu
 
 export type AppMenuProps = {
-    children?: ReactNode
+    leadingItem?: AppMenuItem
+    mainItems?: AppMenuItem[]
+    trailingItem?: AppMenuItem
+    secondaryItems?: AppMenuItem[]
+}
+
+export type AppMenuItem = AppMenuTextItem | AppMenuIconItem
+
+type AppMenuItemBase = {
+    url?: string
+    onClick?: () => void
+    testId?: string
+}
+
+export type AppMenuTextItem = {
+    title: ReactElement
+} & AppMenuItemBase
+
+export function isAppMenuTextItem(item: AppMenuItem): item is AppMenuTextItem {
+    return typeof (item as AppMenuTextItem).title === 'object'
+}
+
+export type AppMenuIconItem = {
+    icon: (props: SVGProps<SVGSVGElement>) => JSX.Element
+    iconClasses?: string
+} & AppMenuItemBase
+
+export function isAppMenuIconItem(item: AppMenuItem): item is AppMenuIconItem {
+    return typeof (item as AppMenuIconItem).icon === 'object'
+}
+
+function MenuElement(props: AppMenuItemProps): JSX.Element {
+    return (
+        <OuterAppMenuElement {...props}>
+            <InnerAppMenuElement {...props} />
+        </OuterAppMenuElement>
+    )
+}
+
+type AppMenuItemProps = {
+    fromItem?: AppMenuItem
```

Comment:
Couldn't this be non-optional? Is there a case where a menu element does not have an item?